File: apps/backend/novel-document-service/src/main/java/com/deusexmachina/novel/document/model/Document.java
package com.deusexmachina.novel.document.model;

import lombok.Builder;
import lombok.Data;
import lombok.NonNull;

import java.time.Instant;
import java.util.List;
import java.util.Map;

@Data
@Builder
public class Document {
    
    @NonNull
    private String id;
    
    @NonNull
    private String contextId;
    
    @NonNull
    private String title;
    
    private String subtitle;
    
    private String authorId;
    
    private String authorName;
    
    @NonNull
    @Builder.Default
    private DocumentStatus status = DocumentStatus.DRAFT;
    
    @NonNull
    @Builder.Default
    private Instant createdAt = Instant.now();
    
    @NonNull
    @Builder.Default
    private Instant updatedAt = Instant.now();
    
    private String description;
    
    private String genre;
    
    private List<String> tags;
    
    @Builder.Default
    private int wordCount = 0;
    
    @Builder.Default
    private int chapterCount = 0;
    
    @Builder.Default
    private int sceneCount = 0;
    
    private String currentVersion;
    
    private List<String> chapterIds;
    
    private DocumentSettings settings;
    
    private Map<String, Object> metadata;
    
    @Builder.Default
    private boolean active = true;
    
    @Data
    @Builder
    public static class DocumentSettings {
        @Builder.Default
        private int targetWordCount = 80000;
        
        @Builder.Default
        private String language = "en";
        
        @Builder.Default
        private String style = "narrative";
        
        @Builder.Default
        private boolean autoSave = true;
        
        @Builder.Default
        private int autoSaveIntervalSeconds = 60;
        
        private Map<String, Object> customSettings;
    }
    
    public void updateWordCount(int newCount) {
        this.wordCount = newCount;
        this.updatedAt = Instant.now();
    }
    
    public void incrementChapterCount() {
        this.chapterCount++;
        this.updatedAt = Instant.now();
    }
    
    public void incrementSceneCount() {
        this.sceneCount++;
        this.updatedAt = Instant.now();
    }
}

File: apps/frontend/novel-creator/src/types/document.ts
export interface Document {
  id: string
  contextId: string
  title: string
  subtitle?: string
  authorId: string
  authorName: string
  status: DocumentStatus
  createdAt: Date
  updatedAt: Date
  description?: string
  genre?: string
  tags?: string[]
  wordCount: number
  chapterCount: number
  sceneCount: number
  currentVersion?: string
  settings?: DocumentSettings
  metadata?: Record<string, unknown>
}

export type DocumentStatus = 'DRAFT' | 'IN_REVIEW' | 'PUBLISHED' | 'ARCHIVED'

export interface DocumentSettings {
  targetWordCount?: number
  language?: string
  style?: string
  autoSave?: boolean
  autoSaveIntervalSeconds?: number
  customSettings?: Record<string, unknown>
}

export interface Chapter {
  chapterNumber: number
  title: string
  summary?: string
  scenes: Scene[]
  wordCount: number
  status: ChapterStatus
}

export interface Scene {
  sceneNumber: number
  title?: string
  content: string
  type: SceneType
  wordCount: number
  characters?: string[]
  location?: string
  timeOfDay?: string
  mood?: string
  metadata?: Record<string, unknown>
}

export type ChapterStatus = 'OUTLINE' | 'DRAFT' | 'REVISION' | 'FINAL'
export type SceneType = 'ACTION' | 'DIALOGUE' | 'DESCRIPTION' | 'FLASHBACK' | 'TRANSITION' | 'CLIMAX'

export interface CreateDocumentInput {
  title: string
  subtitle?: string
  description?: string
  genre?: string
  tags?: string[]
  settings?: DocumentSettings
}